atomicfu/src/appleMain/kotlin/kotlinx/atomicfu/locks/Synchronized.kt

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kotlinx.atomicfu.locks
+
+import kotlin.native.ref.createCleaner
+import kotlinx.atomicfu.*
+import kotlinx.cinterop.Arena
+import kotlinx.cinterop.ExperimentalForeignApi
+import kotlinx.cinterop.alloc
+import kotlinx.cinterop.memScoped
+import kotlinx.cinterop.ptr
+import kotlinx.cinterop.value
+import kotlinx.cinterop.IntVar
+import kotlinx.cinterop.UIntVar
+import kotlinx.cinterop.toCPointer
+import kotlinx.cinterop.toLong
+import platform.posix.pthread_cond_destroy
+import platform.posix.pthread_cond_init
+import platform.posix.pthread_cond_signal
+import platform.posix.pthread_cond_t
+import platform.posix.pthread_cond_wait
+import platform.posix.pthread_get_qos_class_np
+import platform.posix.pthread_mutex_destroy
+import platform.posix.pthread_mutex_init
+import platform.posix.pthread_mutex_lock
+import platform.posix.pthread_mutex_t
+import platform.posix.pthread_mutex_unlock
+import platform.posix.pthread_mutexattr_destroy
+import platform.posix.pthread_mutexattr_init
+import platform.posix.pthread_mutexattr_settype
+import platform.posix.pthread_mutexattr_t
+import platform.posix.pthread_override_qos_class_end_np
+import platform.posix.pthread_override_qos_class_start_np
+import platform.posix.pthread_override_t
+import platform.posix.pthread_self
+import platform.posix.qos_class_self
+
+import platform.posix.PTHREAD_MUTEX_ERRORCHECK
+
+
+@ThreadLocal
+var currentThreadId = 0L
+
+// Based on the compose-multiplatform-core implementation with added qos and the pool back-ported
+// from the atomicfu implementation.
+public actual open class SynchronizedObject {
+
+    companion object {
+        private const val NO_OWNER = 0L
+    }
+
+    private val owner: AtomicLong = atomic(NO_OWNER)
+    private var reEnterCount: Int = 0
+    private val waiters: AtomicInt = atomic(0)
```

Comment:
Calling this `waiters` is semantically confusing: if there exists only a single thread that locks the lock, `waiters` fluctuates between 0 and 1.

Have you considered renaming it? (e.g. `threadsOnLock` or something like that) and dropping a comment?

Reply:
Renamed as suggested